lib/greengrasscoreipc/model.ts

```diff
@@ -0,0 +1,880 @@
+ */
+export enum MetricUnitType {
+
+    BYTES = "BYTES",
+
+    BYTES_PER_SECOND = "BYTES_PER_SECOND",
+
+    COUNT = "COUNT",
+
+    COUNT_PER_SECOND = "COUNT_PER_SECOND",
+
+    MEGABYTES = "MEGABYTES",
+
+    SECONDS = "SECONDS"
+
+}
+
+/**
+ * The context is ignored if used in PublishMessage.
+ */
+export interface MessageContext {
+
+    topic?: string
+
+}
+
+export interface ValidateConfigurationUpdateEvent {
+
+    configuration?: any,
+
+    deploymentId: string
+
+}
+
+export interface RunWithInfo {
+
+    posixUser?: string,
+
+    windowsUser?: string,
+
+    systemResourceLimits?: SystemResourceLimits
+
+}
+
+export interface PreComponentUpdateEvent {
+
+    deploymentId: string,
+
+    isGgcRestarting: boolean
+
+}
+
+export interface PostComponentUpdateEvent {
+
+    deploymentId: string
+
+}
+
+export interface MQTTMessage {
+
+    topicName: string,
+
+    payload?: eventstream.Payload
+
+}
+
+export interface MQTTCredential {
+
+    clientId?: string,
+
+    certificatePem?: string,
+
+    username?: string,
+
+    password?: string
+
+}
+
+export interface Metric {
+
+    name: string,
+
+    unit: MetricUnitType,
+
+    value: number
+
+}
+
+/**
+ * To preserve backwards compatibility, no validation is performed on enum-valued fields.
+ */
+export enum LifecycleState {
+
+    RUNNING = "RUNNING",
+
+    ERRORED = "ERRORED",
+
+    NEW = "NEW",
+
+    FINISHED = "FINISHED",
+
+    INSTALLED = "INSTALLED",
+
+    BROKEN = "BROKEN",
+
+    STARTING = "STARTING",
+
+    STOPPING = "STOPPING"
+
+}
+
+export interface JsonMessage {
+
+    message?: any,
+
+    context?: MessageContext
+
+}
+
+/**
+ * To preserve backwards compatibility, no validation is performed on enum-valued fields.
+ */
+export enum DeploymentStatus {
+
+    QUEUED = "QUEUED",
+
+    IN_PROGRESS = "IN_PROGRESS",
+
+    SUCCEEDED = "SUCCEEDED",
+
+    FAILED = "FAILED"
+
+}
+
+/**
+ * To preserve backwards compatibility, no validation is performed on enum-valued fields.
+ */
+export enum ConfigurationValidityStatus {
+
+    ACCEPTED = "ACCEPTED",
+
+    REJECTED = "REJECTED"
+
+}
+
+export interface ConfigurationUpdateEvent {
+
+    componentName: string,
+
+    keyPath: string[]
+
+}
+
+export interface CertificateUpdate {
+
+    privateKey?: string,
+
+    publicKey?: string,
+
+    certificate?: string,
+
+    caCertificates?: string[]
+
+}
+
+/**
+ * To preserve backwards compatibility, no validation is performed on enum-valued fields.
+ */
+export enum CertificateType {
+
+    SERVER = "SERVER"
+
+}
+
+export interface BinaryMessage {
+
+    message?: eventstream.Payload,
+
+    context?: MessageContext
+
+}
+
+export interface ValidateConfigurationUpdateEvents {
+
+    validateConfigurationUpdateEvent?: ValidateConfigurationUpdateEvent
+
+}
+
+export interface SubscriptionResponseMessage {
+
+    jsonMessage?: JsonMessage,
+
+    binaryMessage?: BinaryMessage
+
+}
+
+export interface SecretValue {
```

Comment:
this is from a union, other language implementations have it where there's a enum that tells you which variant of the union is set. Not sure if we should replicate that here or not.

Reply:
This could be a little tricky.  The model is intentionally 100% POD types and it wouldn't be appropriate to inject unmodeled data into modeled shapes.  One possibility is to have the "selector" enums live in the model and add free functions for each union type that take the union as input and return the appropriate selector enum as output.  I feel like that doesn't naturally have great visibility or an obvious API contract though.